### PR TITLE
#3 Elo, PvP performance matrix

### PR DESCRIPTION
**Pull request by donren-leung**
## Main
- Elo ratings! And H2H summary matrix output -- see [below](#Output-V2)
  - allows seeing whether a bot has any weaknesses. \
    For example, ScaredyCat (1830) underperforms `PR:1665 (-Δ)` against CopyCat (1364).
- `BotTracker` class
  - tracks a `Bot`, keeps `GameResultStat` (moved from controller.py), and
  - calculates Elo rating using variable K-factor based on games played.

## Other
 - fixed pre-commit hook.
 - `Game.run()` refactored to return `GameResult` in prep for multiprocessing.

## Output V2
```
Head-to-Head Statistics
====================================================================================================

                          [1] FirstMover (1281)          [2] GreedyCat (1704)         [3] ScaredyCat (1830)           [4] CopyCat (1364)      
----------------------------------------------------------------------------------------------------------------------------------------------
[0] RandomBot (1321)      11/8/11  [0] PR:1281 (-Δ)      2/3/25  [0] PR:1353 (+Δ)      2/0/28  [0] PR:1372 (+Δ)     14/0/16  [0] PR:1340 (+Δ) 
                        Δ(PR):±40  [1] PR:1321 (+Δ)   Δ(PR):±32  [2] PR:1673 (-Δ)   Δ(PR):±51  [3] PR:1780 (-Δ)   Δ(PR):±19  [4] PR:1344 (-Δ) 

[1] FirstMover (1281)  xxxxxxxxxxxxxxxxxxxxxxxxxxxx      0/0/30  [1] PR:  -∞ (-Δ)      0/0/30  [1] PR:  -∞ (-Δ)     8/11/11  [1] PR:1329 (+Δ) 
                       xxxxxxxxxxxxxxxxxxxxxxxxxxxx    Δ(PR):±∞  [2] PR:   ∞ (+Δ)    Δ(PR):±∞  [3] PR:   ∞ (+Δ)   Δ(PR):±48  [4] PR:1316 (-Δ) 

[2] GreedyCat (1704)   ----------------------------  xxxxxxxxxxxxxxxxxxxxxxxxxxxx     0/15/15  [2] PR:1640 (-Δ)      23/6/1  [2] PR:1689 (-Δ) 
                       ----------------------------  xxxxxxxxxxxxxxxxxxxxxxxxxxxx   Δ(PR):±65  [3] PR:1895 (+Δ)   Δ(PR):±15  [4] PR:1379 (+Δ) 

[3] ScaredyCat (1830)  ----------------------------  ----------------------------  xxxxxxxxxxxxxxxxxxxxxxxxxxxx      21/9/0  [3] PR:1665 (-Δ) 
                       ----------------------------  ----------------------------  xxxxxxxxxxxxxxxxxxxxxxxxxxxx  Δ(PR):±166  [4] PR:1529 (+Δ) 

[4] CopyCat (1364)     ----------------------------  ----------------------------  ----------------------------  xxxxxxxxxxxxxxxxxxxxxxxxxxxx 
                       ----------------------------  ----------------------------  ----------------------------  xxxxxxxxxxxxxxxxxxxxxxxxxxxx 


====================================================================================================
```

## Output V1
```
Head-to-Head Statistics
====================================================================================================

                           [0] RandomBot (1266)         [1] FirstMover (1290)          [2] GreedyCat (1725)         [3] ScaredyCat (1878)           [4] CopyCat (1340)      
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[0] RandomBot (1266)    xxxxxxxxxxxxxxxxxxxxxxxxxx       4/5/11  PR=1163 (Δ=-103)      0/2/18  PR=1214 (Δ=-52)       1/0/19  PR=1367 (Δ=+101)     10/0/10  PR=1340 (Δ=+75)  
                        xxxxxxxxxxxxxxxxxxxxxxxxxx               PR=1393 (Δ=+103)              PR=1777 (Δ=+52)               PR=1777 (Δ=-101)              PR=1266 (Δ=-75)  
[1] FirstMover (1290)   --------------------------    xxxxxxxxxxxxxxxxxxxxxxxxxx       0/0/20  PR=925 (Δ=-365)       0/0/20  PR=1078 (Δ=-212)       3/9/8  PR=1252 (Δ=-39)  
                        --------------------------    xxxxxxxxxxxxxxxxxxxxxxxxxx               PR=2090 (Δ=+365)              PR=2090 (Δ=+212)              PR=1379 (Δ=+39)  
[2] GreedyCat (1725)    --------------------------    --------------------------    xxxxxxxxxxxxxxxxxxxxxxxxxx      0/10/10  PR=1687 (Δ=-38)       15/5/0  PR=1678 (Δ=-47)  
                        --------------------------    --------------------------    xxxxxxxxxxxxxxxxxxxxxxxxxx               PR=1916 (Δ=+38)               PR=1387 (Δ=+47)  
[3] ScaredyCat (1878)   --------------------------    --------------------------    --------------------------    xxxxxxxxxxxxxxxxxxxxxxxxxx       18/2/0  PR=1852 (Δ=-26)  
                        --------------------------    --------------------------    --------------------------    xxxxxxxxxxxxxxxxxxxxxxxxxx               PR=1367 (Δ=+26)  
[4] CopyCat (1340)      --------------------------    --------------------------    --------------------------    --------------------------    xxxxxxxxxxxxxxxxxxxxxxxxxx  
                        --------------------------    --------------------------    --------------------------    --------------------------    xxxxxxxxxxxxxxxxxxxxxxxxxx  

====================================================================================================
```

### CODE REVIEW

**Comment by donren-leung** [comment]
Nitpick, `Δ(PR)` in summary is actually the difference between the player's rating and performance rating (it's the same but negative for the other player). I think currently it could get confused as the difference between the two PRs which it isn't.

Any ideas for a better name?

File: checkers_bot_tournament/checkers_util.py [hunk 1]
```diff
@@ -20,8 +26,53 @@ def make_unique_bot_string(*args, **_kwargs) -> str:
...
         idx, bot_str = args
         return f"[{idx}] {bot_str}"
     else:
         raise TypeError("Invalid arguments to make_unique_bot_string")
+
+
+# Function to compute Performance Rating of player against a specific
+# opponent given both Elo ratings
+def compute_performance_rating(w: int, d: int, l: int, bot_rating: float, opp_rating: float):
+    # https://en.wikipedia.org/wiki/Performance_rating_(chess)
```
Comment (trentzz) [inline checkers_bot_tournament/checkers_util.py]: Could make this a `""" """` docstring instead?

File: checkers_bot_tournament/bots/bot_tracker.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    @property
+    def total_losses(self):
+        return self.white_losses + self.black_losses
+
+    @property
+    def total_games(self):
+        return self.total_wins + self.total_draws + self.total_losses
+
+
+STARTING_ELO = 1500
```
Comment (trentzz) [inline checkers_bot_tournament/bots/bot_tracker.py]: can we move these static values somewhere else? global static variables are a bit sussy, maybe just a class at the top of this file like:

```python
class EloConfig:
    STARTING_ELO = 1500
    SCALE = 400
    
EloConfig.STARTING_ELO
```

File: checkers_bot_tournament/stat_printing.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                f"{draw_str:<{col_width}}"
+                f"{loss_str:<{col_width}}"
+                f"= {score_str}"
+            )
+            file.write(pct_str + "\n")
+
+        file.write("=" * 60 + "\n\n")
+
+
+def write_tournament_h2h_stats(bot_list: list[BotTracker], file: IO) -> None:
```
Comment (trentzz) [inline checkers_bot_tournament/stat_printing.py:77]: not that important but this do be a bit of a chonky function idk if it's worth splitting up

**Comment by donren-leung** [comment]
Ready when you are!

**Comment by trentzz** [comment]
Okayy merging :D

**Pull request merged**